### ISSUE TITLE
invalid free when $SIG{__WARN__} is set during INIT or END

### ISSUE THREAD

**Issue by toddr** [rurban/perl-compiler#306]
```
$>perl -e 'INIT { $SIG{__WARN__} = sub { die } } print "prepare to die\n";'
prepare to die
```

```
$>perlcc -e 'INIT { $SIG{__WARN__} = sub { die } } print "prepare to die\n";'; ./a.out
prepare to die
*** glibc detected *** ./a.out: free(): invalid pointer: 0x0804ea80 ***

$ gdb ./a.out

gdb) c
Continuing.
prepare to die
ccode3060(48044,0x7fff77ab2300) malloc: *** error for object 0x1003c6c48: pointer being freed was not allocated
*** set a breakpoint in malloc_error_break to debug

Program received signal SIGABRT, Aborted.
0x00007fff9694a286 in ?? () from /usr/lib/system/libsystem_kernel.dylib
(gdb) bt
#0  0x00007fff9694a286 in ?? () from /usr/lib/system/libsystem_kernel.dylib
#1  0x00007fff8b0569f9 in pthread_kill () from /usr/lib/system/libsystem_pthread.dylib
#2  0x00007fff956379b3 in abort () from /usr/lib/system/libsystem_c.dylib
#3  0x00007fff8743f1cb in free () from /usr/lib/system/libsystem_malloc.dylib
#4  0x0000000100246e5a in Perl_sv_free2 (sv=0x10046eaa8 <sv_list+392>, rc=1) at sv.c:7035
#5  0x0000000100029e4f in S_SvREFCNT_dec (sv=0x10046eaa8 <sv_list+392>) at inline.h:166
#6  0x000000010002cbf9 in Perl_op_clear (o=0x10046e100 <unopaux_list>) at op.c:1049
#7  0x000000010002c5b0 in Perl_op_free (o=0x10046e100 <unopaux_list>) at op.c:786
#8  0x0000000100106666 in Perl_cv_undef_flags (cv=0x10046ea70 <sv_list+336>, flags=0) at pad.c:340
#9  0x00000001001061e0 in Perl_cv_undef (cv=0x10046ea70 <sv_list+336>) at pad.c:301
#10 0x0000000100243a9e in Perl_sv_clear (orig_sv=0x10046ea70 <sv_list+336>) at sv.c:6602
#11 0x0000000100246d86 in Perl_sv_free2 (sv=0x10046ea70 <sv_list+336>, rc=1) at sv.c:7033
#12 0x00000001002aa231 in S_SvREFCNT_dec (sv=0x10046ea70 <sv_list+336>) at inline.h:166
#13 0x00000001002b245e in Perl_leave_scope (base=0) at scope.c:973
#14 0x00000001000868a5 in S_my_exit_jump () at perl.c:5043
#15 0x0000000100086582 in Perl_my_exit (status=0) at perl.c:4937
#16 0x00000001000782e9 in S_run_body (oldscope=1) at perl.c:2450
#17 0x0000000100077660 in perl_run (my_perl=0x1007001b0) at perl.c:2371
#18 0x00000001000056c4 in main (argc=1, argv=0x7fff5fbff660, env=0x7fff5fbff670) at ccode3060.c:806
(gdb) f 4
#4  0x0000000100246e5a in Perl_sv_free2 (sv=0x10046eaa8 <sv_list+392>, rc=1) at sv.c:7035
7035	            del_SV(sv);
(gdb) p *sv
$1 = {sv_any = 0x10046e160 <xpv_list+32>, sv_refcnt = 0, sv_flags = 255, sv_u = {
    svu_pv = 0x100716380 "__WARN__", svu_iv = 4302398336, svu_uv = 4302398336,
    svu_nv = 2.1256672125421445e-314, svu_rv = 0x100716380, svu_rx = 0x100716380, svu_array = 0x100716380,
    svu_hash = 0x100716380, svu_gp = 0x100716380, svu_fp = 0x100716380}, sv_debug_optype = 0,
  sv_debug_inpad = 0, sv_debug_line = 0, sv_debug_serial = 0,
  sv_debug_file = 0x1003c6c48 "__WARN__ sv_list[7] 0x10004403", sv_debug_parent = 0x0}
```

So it fails when destructing the __WARN__ key in the INIT cv.
The probable fix is to bump refcounts of init_av and end_av (the INIT and END blocks CVs).

**Comment by rurban** [comment]
High? I won't make all warnings dynamic just to please this usecase.

**Comment by toddr** [comment]
High because simply setting a signal in a begin/end block is enough to cause a segfault. I don't follow what this has to do with warnings?

**Comment by toddr** [comment]
If you prefer...

perlcc -e 'INIT { $SIG{__DIE__} = sub { die } } print "prepare to die\n";'; ./a.out

**Comment by toddr** [comment]
It's also entirely possible this is a dupe of #271

**Comment by rurban** [comment]
Fixed with f939b7fc117a5c396bc0ddc67522d07877056948

**Comment by toddr** [comment]
This may fix #271

**Comment by toddr** [comment]
Our commit looks like this and appears sufficient.

https://github.com/CpanelInc/perl-compiler/commit/08bd4545aeb3f9ec38efe657777670c6c2b72353

**Comment by rurban** [comment]
Fixed and added testcase with commit 3b17627a85d27a0b10a65ff7813d2c255f60dcb3
Author: Reini Urban <rurban@cpanel.net>
Date:   Mon Nov 2 19:26:33 2015 +0100

    C 1.52_12: defer INIT{} and END{} destruction

    bump the refcount of INIT{} and END{} CVs. See GH #306.
    Ignore destruction problems on those CVs.
